### İki Popülasyon Ortalama Farkı Güven Aralığı
Her iki popülasyonun da varyansı da biliniyorsa Z dağılım, iki popülasyonun varyansı bilinmiyorsa ve gözlem sayısı (n<30) ise t dağılımını kullanacağız ancak t dağılımının da gözlem sayısı arttıkça z dağılımına yakınsadığını unutmamak gerekir yani büyük gözlem sayılarında popülasyon varyansı bilinmese bile z dağılımı kullanılabilir.

- Şimdiye kadar tek bir popülasyonun ortalamasını tahmin ediyorduk. Bu konuda iki farklı grubun ortalamaları arasındaki farkın güven aralığını hesaplıyoruz örneğin "A grubunun ortalaması, B grubunun ortalamasından ne kadar farklı, ve bu fark ne kadar güvenilir?"
- Eğer bulduğumuz sonuç 0'ı içeriyorsa her iki grup arasında anlamlı bir istatistiksel fark yoktur sonucuna varabiliriz çünkü 0 güven aralığının içinde kalır.

### Bu kısımda 2 farklı duruma göre 2 farklı formül kullanacağız.

- 1. Formül: Eğer iki grubun varyanslarının birbirine eşit olduğu varsayılıyorsa , iki grubun standart sapmaları birleştirilerek ortak bir varyans hesaplanır.
$$(\bar{x}_1 - \bar{x}_2) \pm t_{\alpha/2, \, df} \cdot \sqrt{s_p^2 \left(\frac{1}{n_1} + \frac{1}{n_2}\right)}$$

Ortak varyans ($s_p^2$): $$s_p^2 = \frac{(n_1 - 1)s_1^2 + (n_2 - 1)s_2^2}{n_1 + n_2 - 2}$$

Serbestlik Derecesi ($df$): $$df = n_1 + n_2 - 2$$


- 2. Formül: Eğer iki grubun varyanslarının eşit olmadığı varsayılıyorsa, gruplar birleştirilmez. Bu durumda serbestlik derecesini bulmak için daha karmaşık bir formül (Welch-Satterthwaite) kullanılır.
$$(\bar{x}_1 - \bar{x}_2) \pm t_{\alpha/2, \, df} \cdot \sqrt{\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}}$$

Serbestlik Derecesi ($df$) hesaplaması:
$$df = \frac{\left(\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}\right)^2}{\frac{\left(\frac{s_1^2}{n_1}\right)^2}{n_1 - 1} + \frac{\left(\frac{s_2^2}{n_2}\right)^2}{n_2 - 1}}$$




In [1]:
import numpy as np
from math import sqrt
from scipy import stats
print("******Varyanslar Birbirine Eşit Olduğunu Varsayalım.******")
grup_A = np.array([78, 82, 85, 91, 76, 88, 95, 79, 84, 90])   # Örnek: A yöntemiyle eğitim alan grubun sınav notları
grup_B = np.array([65, 70, 72, 68, 75, 71, 69, 73, 66, 74])   # Örnek: B yöntemiyle eğitim alan grubun sınav notları
grup_a_ortalama = np.mean(grup_A)
grup_b_ortalama = np.mean(grup_B)
grup_a_sapma = np.std(grup_A, ddof=1)
grup_b_sapma = np.std(grup_B, ddof=1)
ortalamalar_farki = grup_a_ortalama - grup_b_ortalama
t_alt_sinir = stats.t.ppf(0.025, df = (len(grup_A) + len(grup_B) - 2))
t_ust_sinir = stats.t.ppf(0.975, df = (len(grup_A) + len(grup_B) - 2))
ortak_varyans = ((len(grup_A) - 1) * (grup_a_sapma ** 2) + (len(grup_B) - 1) * (grup_b_sapma ** 2)) / (len(grup_A)+len(grup_B) - 2)
guven_araligi_alt_sinir = (ortalamalar_farki) + (t_alt_sinir * sqrt(ortak_varyans * ((1/len(grup_A))+(1/len(grup_B)))))
guven_araligi_ust_sinir = (ortalamalar_farki) + (t_ust_sinir * sqrt(ortak_varyans * ((1/len(grup_A))+(1/len(grup_B)))))
print(f"Ortalamalar farkı: {ortalamalar_farki}")
print(f"%95 güven aralığı: {guven_araligi_alt_sinir} ile {guven_araligi_ust_sinir}")

******Varyanslar Birbirine Eşit Olduğunu Varsayalım.******
Ortalamalar farkı: 14.5
%95 güven aralığı: 9.824694445013156 ile 19.175305554986842


In [ ]:
print("******Varyanslar Birbirine Eşit Olmadığını Varsayalım.******")
ortalamalar_farki = grup_a_ortalama - grup_b_ortalama
grup_a_sapma = np.std(grup_A, ddof=1)
grup_b_sapma = np.std(grup_B, ddof=1)
serbestlik_derecesi = serbestlik_derecesi = ((((grup_a_sapma ** 2) / len(grup_A)) + ((grup_b_sapma ** 2) / len(grup_B))) ** 2) / (
    (((grup_a_sapma**2/len(grup_A))**2) / (len(grup_A) - 1)) 
    + (((grup_b_sapma**2/len(grup_B))**2) / (len(grup_B) - 1))
)
standart_hata = sqrt(((grup_a_sapma ** 2) / len(grup_A)) + ((grup_b_sapma ** 2) / len(grup_B)))
alt_sinir = ortalamalar_farki + (stats.t.ppf(0.025, serbestlik_derecesi) * standart_hata)
ust_sinir = ortalamalar_farki + (stats.t.ppf(0.975, serbestlik_derecesi) * standart_hata)
print(f"%95 Güven düzeyiyle aralık {alt_sinir} ile {ust_sinir} arasındadır.")

******Varyanslar Birbirine Eşit Olmadığını Varsayalım.******
%95 Güven düzeyiyle aralık 9.720912544852956 ile 19.279087455147042 arasındadır.


In [6]:
from scipy.stats import ttest_ind
sonuc = ttest_ind(grup_A, grup_B, equal_var=False)
print(f"SciPy Serbestlik Derecesi df: {sonuc.df}")
print(f"Elle Hesapladığımız Serbestlik Derecesi df: {serbestlik_derecesi}")

SciPy Serbestlik Derecesi df: 13.810019390639702
Elle Hesapladığımız Serbestlik Derecesi df: 13.810019390639708


### Sonuç
SciPy'ın hesapladığı serbestlik derecesi (13.81) ile elle hesapladığımız serbestlik derecesi (13.81) birebir örtüşüyor, bu da Welch formülünü doğru uyguladığımızı doğruluyor.

İlginç bir nokta: Grup A'nın standart sapması (≈6.3) ile Grup B'nin standart sapması (≈3.2) birbirinden oldukça farklı olmasına rağmen, Pooled yöntemiyle bulduğumuz güven aralığı (9.82, 19.18) ile Welch yöntemiyle bulduğumuz güven aralığı (9.72, 19.28) birbirine çok yakın çıktı. Bunun sebebi, iki grubun gözlem sayılarının (n=10, n=10) eşit olması varyanslar eşit olmasa da, eşit örneklem büyüklüğü bu farkın etkisini azaltıyor. Varyans farkı çok daha büyük olsaydı veya örneklem büyüklükleri (n) birbirinden farklı olsaydı, 
Pooled ve Welch sonuçları arasındaki fark çok daha belirgin olurdu bu durumda Welch yöntemi daha güvenilir sonuç verir.

Bu yüzden pratikte, varyansların eşit olup olmadığından emin değilsek, Welch yöntemini varsayılan olarak tercih etmek daha güvenli bir yaklaşımdır.